enhancements/authentication/oauth-proxy-share-configuration.md

```diff
@@ -0,0 +1,98 @@
+---
+title: oauth-proxy-shared-configuration
+authors:
+  - "@deads2k"
+reviewers:
+  - "@standa"
+approvers:
+  - "@sttts"
+creation-date: 2020-05-26
+last-updated: 2020-05-26
+status: provisional|implementable|implemented|deferred|rejected|withdrawn|replaced
+see-also:
+  - https://github.com/openshift/enhancements/pull/22  
+replaces:
+superseded-by:
+---
+
+# oauth-proxy Shared Configuration
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Open Questions [optional]
+
+## Summary
+
+The [oauth-proxy](https://github.com/openshift/oauth-proxy) is a commonly used proxy which allows a naive application to 
+make use of in-cluster authentication and authorization to protect itself on an openshift platform without changing
+the application itself.
+This effectively makes the application appear as an extension to our platform, examples include: logging kibana, 
+monitoring grafana, etc.
+There are a few configuration values that are considered as related to platform configuration or security that the oauth-proxy
+could auto-detect.
+
+## Motivation
+
+If a customer wants to set a value on all oauth-proxy instances such as proxy configuration or `--cookie-expire` duration,
+there is no unified way to set that across all instances of the oauth-proxy.
+Having cluster-admins know every application using the oauth-proxy and figuring out how to configure them all seems 
+abusive.
+
+### Goals
+
+1. Make it easy to describe configuration for all oauth-proxy binaries.
+
+### Non-Goals
+
+
+## Proposal
+
+We will create an oauthproxies.config.openshift.io to ensure that read access can be widely granted without any risk of
+accidentally allowing reading of more private data (versus attaching to oauth.config.openshift.io for instance).
+Its .spec stanza contain typed fields for certain flags.
+
+The following flags for oauth-proxy are good targets for being consistent for all oauth-proxy instances in the cluster.
+If the oauth-proxy does not have one of these flags explicitly set, it will attempt to read the value from oauthproxies.config.openshift.io.
+1. -cookie-expire
```

Comment:
As far as I understand this setting is then orthogonal to the global access token expiry as per https://docs.openshift.com/container-platform/4.4/authentication/configuring-internal-oauth.html#oauth-configuring-internal-oauth_configuring-internal-oauth, correct?

Does a potential discrepancy with the cookie session timeout and the access token expiry have any ramifications consuming applications have to consider?

Reply:
+1 to make this parameter configurable per oauth-proxy instance. It only affects to the cookie time in the browser and it doesn't necesarily has to be aligned with the validity of the token provided in the CLI or any other web based application which also uses oauth-proxy.

Reply:
I didn't mean to express a wish for per-instance configuration but rather if a discrepancy between a global `cookie-expire` timeout and a global access token expiry is a problem at all.

i see the following cases:
- cookie expiry is shorter than access token expiry: in that case a re-login will be enforced as the initial (cookie-encrypted) access token is lost.
- cookie expiry is longer than access token expiry: in that case a re-login will also be enforced as the access token expired.

I.e. from a user perspective the actual "session timeout" is `min(cookie-expiry, access token expiry)` correct?

Reply:
The enhancement does not block per-instance configuration.

> cookie expiry is shorter than access token expiry: in that case a re-login will be enforced as the initial (cookie-encrypted) access token is lost.

Yes.

> cookie expiry is longer than access token expiry: in that case a re-login will also be enforced as the access token expired.

No. The session does not currently expire when the token does as far as I am aware.

Reply:
> No. The session does not currently expire when the token does as far as I am aware.

This seems counter-intuitive to me, do we want to change the current behaviour to match suggestion from @s-urbaniak?

> from a user perspective the actual "session timeout" is min(cookie-expiry, access token expiry)

Reply:
I'd still consider some override method per instance so different instances cna have different timeouts without requiring editing each operator or each deployment.

Let's say we can have an override per namespace and/or a matchLabel selector.

What I mean, no matter the implementation, is that it would be desirable to have configurable (and why not different) values for token/cookie timeout.

Reply:
> This seems counter-intuitive to me, do we want to change the current behaviour to match suggestion from @s-urbaniak?

currently not possible without further development investment into the oauth-proxy

> I'd still consider some override method per instance so different instances cna have different timeouts without requiring editing each operator or each deployment.

use options of the binary